migrate_legacy_captures.js

```diff
@@ -0,0 +1,218 @@
+/**
+ *  This is a migration script that is to get all the records from legacy `public.tree` and populate
+ *  table `treetracker.capture` as long as the legacy records in `public.trees` are approved,
+ *  active, has a valid image url and if legacy data has not made it to capture table already.
+ *  In addition, if the concept of planter referenced in legacy trees table is not found in the
+ *  `treetracker.grower_account`, it is created first using the data from `public.planter` before proceeding
+ *  with populating the `treetracker.captures` table.
+ *
+ *  @created Jan 28 2022
+ *
+ * */
+
+require("dotenv").config();
+const generate_uuid_v4 = require("uuid").v4;
+
+const { knex } = require("./database/knex");
+const Writable = require("stream").Writable;
+const ws = Writable({ objectMode: true });
+const ProgressBar = require("progress");
+
+async function ensureGrowerAccountExists(legacyTree, transaction) {
+  let planter = null;
+  if (!legacyTree.planter_identifier && !legacyTree.planter_id) {
+    throw Error("No planter record found in legacy tree table");
+  }
+  if (legacyTree.planter_id) {
+    planter = await transaction
+      .select()
+      .table("public.planter")
+      .where("id", legacyTree.planter_id)
+      .first();
+  } else if (
+    legacyTree.planter_identifier &&
+    legacyTree.planter_identifier.trim()
+  ) {
+    planter = await transaction
+      .select()
+      .table("public.planter")
+      .where("email", legacyTree.planter_identifier)
+      .orWhere("phone", legacyTree.planter_identifier)
+      .first();
+  }
+  if (!planter) {
+    throw Error(
+      `No planter record found for tree entry with id ${legacyTree.id}`
+    );
+  }
+  let growerAccount = await transaction("treetracker.grower_account")
+    .where("email", planter.email)
+    .orWhere("phone", planter.phone)
+    .first();
+  if (!growerAccount) {
+    console.log(
+      `GrowerAccount not found for planter_identifier ${legacyTree.planter_identifier}`
+    );
+    console.log(
+      `Populating grower account for ${legacyTree.planter_identifier}`
+    );
+    // status, first_registration_at aren't available in the source `public.planter` table
+    // Is it ok, if the above attributes take the db defaults?
+    growerAccount = await transaction("treetracker.grower_account").insert(
+      {
+        name: planter.first_name + " " + planter.last_name,
+        email: planter.email,
+        phone: planter.phone,
+        image_url: planter.image_url ?? legacyTree.planter_photo_url, //should this be public.trees.planter_photo_url?
+        image_rotation: planter.image_rotation ?? 0,
+        organization_id: planter.organization_id,
+        wallet_id: generate_uuid_v4(), // TODO: temporary workaround, needs to be a specific uuid and not a newly generated one.
+        wallet: "Temporary placeholder",
+        first_registration_at: new Date(),

```

Comment:
The source table `public.trees` doesn't have created_at column or any date column. Not sure what would be a good strategy here. We could look up the oldest created_at entry in `public.trees` for the given planter to get the approximate date of the first activity by the planter to populate the `first_registration_at`. But it would be another step in the migration. Wanted to check if this warrants that additional step.

Reply:
planter_registrations.created_at is the column to use.  Basically every planter record will have one or more planter_registration records.  The oldest one is the first registration.